README
README For BloGTK Weblog Client
Version 2.0
"Broadsheet"

http://blogtk.jayreding.com/

CONTENTS:
---------
1. ABOUT BLOGTK
2. SUPPORTED WEBLOGS
3. REQUIRED LIBRARIES
4. INSTALLATION
5. UPGRADING FROM PREVIOUS VERSIONS
6. USING BLOGTK
  6.1  CONFIGURING ACCOUNTS
    6.1.1  CONFIGURING WORDPRESS
    6.1.2  CONFIGURING MOVABLE TYPE/MELODY
    6.1.3  CONFIGURING WINDOWS LIVE SPACES	
  6.2  USING THE BLOG MANAGER
  6.3  USING THE POST EDITOR
7. RELEASE NOTES
8. PROGRAM CREDITS

1. ABOUT BLOGTK:

BloGTK is a weblog posting client written in Python for any platform that 
supports Python and the GTK widget set. BloGTK allows you to connect to your 
weblog with XML-RPC and create new posts.

2. SUPPORTED WEBLOGS:

 - MovableType versions 2.5 or later.
 - Blogger
 - WordPress 1.5 or later or WordPress.com
 - Windows Live Spaces
 - Expression Engine (but autoconfiguration does not work)
 - Any blog that uses the MetaWeblog API.

3. REQUIRED LIBRARIES:

BloGTK requires the following programs/libraries.

- Python 2.3 or higher (but BloGTK is not compatible with Python 3.0)
- PyGTK
- python-gnome2
- python-gnome2-extras
- python-gdata
- python-webkit
- python-feedparser

BloGTK will run under other environments (such as KDE) so long as all the 
necessary libraries are installed.

4. INSTALLATION:

BloGTK needs to be installed as the root user.

Unpack and  install the BloGTK tarball using these commands:

   'tar -xjvf <packagename>.tar.bz2'

Switch to the newly created directory and type 'make install' as root and the 
Makefile will copy the files to the correct location in your system. You do not 
need to configure or compile BloGTK as it is written in interpreted Python code.

5. UPGRADING FROM PREVIOUS VERSIONS

BloGTK 2.0 uses an entirely new configuration system, and will not carry over
your prior settings. You will have to re-enter your settings. Your old
settings will remain should you downgrade to an earlier version of BloGTK, not
that you would want to...

6. USING BLOGTK:

BloGTK 2.0 has two main screens, the blog management screen, and the post
editor. The blog management screen lets you see all your posts, edit posts,
delete posts, and view your blog's home page. The editor allows you to edit,
open, and save individual posts.

6.1  CONFIGURING ACCOUNTS

BloGTK will open up on the Settings control panel first. First, type the name
you wish to use for your blog account and the address of your blog's home
page in the top two fields.

BloGTK will use RSD autodiscovery to discover your blog's settings if your blog 
system and template support it. Use the "Autodetect Settings" button to try
autodiscovery. With luck, you will get your endpoint, blog ID, and blogging
system fields filled automatically. If not, you will need to fill them in
manually based on your blogging system's instructions.

You will need to fill out the username and password for your blogs, these
will not be discovered automatically.

You can specify to retrieve up to 50 posts per blog. (For Blogger, you cannot
change the number of posts to retrieve, you will always pull the latest 25.)
This option is set by the "Get up to X Posts per Refresh" setting.

6.1.1  CONFIGURING WORDPRESS

WordPress requires you to enable XML-RPC before using an external editor like
BloGTK. To do this, log in to your blog with the admin account, and go to 
Settings, and then Writing. Check the box next to “XML-RPC” on the Writing 
options page, and save the settings.

6.1.2  CONFIGURING MOVABLE TYPE/MELODY

Movable Type and Melody use a different password than your normal blog login
password. To find this password, log into your blog control panel, then
click your user name in the top right-hand corner. Under "Preferences" you
will find an entry called "Web Services Password". Hit "Reveal" next to the
entry to see the password. This is the password you must use with BloGTK.

6.1.3  CONFIGURING WINDOWS LIVE SPACES

Set your password as per these instructions: 
http://msdn.microsoft.com/en-us/library/bb259698.aspx

That password will be the password used in BloGTK.

6.2  USING THE BLOG MANAGER

BloGTK's main screen is the blog manager. On the left of the screen will
be the list of your blogs. On the top right will be the list of your posts
and below that will be a post preview window. To get your posts, select
the blog you want, and hit the "Refresh" button on the toolbar. BloGTK
will then download your posts.

To edit a post, select a post in the right screen and hit the edit button.

To delete a post, select the post in the right screen and hit the delete
button. You will be asked to confirm your deletion. Once a post is deleted
in cannot be recovered.

To create a new post, select the blog you wish to create a new post for,
and hit the new post button.

The "Home" toolbar button will take you to the homepage of your blog.

6.3  USING THE EDITOR

BloGTK's post editor features built-in syntax highlighting and
spellcheck. The top toolbar allows you to create a new post, open a 
saved post file, save the post, perform cut/copy/paste operations,
undo and redo your last text entries, or post the entry to your blog.

BloGTK saves its posts to disk as XML that can be read in other
applications. The file extension for BloGTK saved post files is 
'.blogtk'.

The second toolbar contains buttons that insert common HTML tags into
your post. You can bold or italicize text, change the alignment of a
text block, add a paragraph or a blockquote, create a list, or add
a link.

On the left is the main entry screen. On the top is the post title
entry field. Below that are tabs corresponding to your main entry,
any extended entry (if your blogging system supports it), and a
preview of your post contents.

Below that is an entry for any tags you may wish to add to your post
(if your blogging system supporst them). Separate each tag by a
comma.

To the right is a selection box for categories, should your blogging
system support them. You may select multiple categories.

Below that are option to allow comments and trackbacks, should your
blogging system support those options. You may save a post as a
draft in your blogging system by selecting the checkbox on the 
bottom right of the editor window.

Under the Edit menu, you may change the date and time of a post,
if your blogging system supports it. You may select the date with
the calendar widget on the left, and the time with the spinbuttons
on the right. BloGTK uses the 24-hour time system by default. The
timestamp should be set to your local time by your blog software,
but some blogging packages will set the time to GMT.

7. RELEASE NOTES:

Bug reports can be filed through the BloGTK project on Launchpad, at
https://launchpad.net/blogtk

Donations for this software would be greatly appreciated. This software is 
totally free, but sending a few bucks to jay@jayreding.com via PayPal would 
bring you good karma, and it'll also mean you'll get extra Developer Suck-Up 
Club support. Plus, it will ensure that I have enough caffeine to develop new 
versions. You can now also donate to the BloGTK project through the BloGTK 
homepage on Sourceforge.

This program was developed under Ubuntu Linux and is released under the
Apache 2.0 license.

8. PROGRAM CREDITS:

Special thanks to James Henstridge for the excellent Python-GTK bindings and 
their equally excellent documentation.

Thanks to Ethan Zimmerman for his bug reports and the original BloGTK icon set.

Thanks to all BloGTK users for bug reports, feature requests, and their support 
of this application.


ChangeLog
blogtk (2.0~ppa4~karmic) karmic; urgency=low

  * Fixed major bug with WebKit preview. (LP: #435593)
  * Fixed bug with close dialog not fuctioning properly in editor.
  * Fixed bug preventing font selection dialog from openining.
  * Confirmed bug fix for Fedora crashing bug. (LP: #435540)

 -- Jay Reding <wombat@serenity>  Thu, 19 Nov 2009 14:51:35 -0600

blogtk (2.0~ppa3~jaunty) jaunty; urgency=low

  * Editor preview no longer autoupdates when title changes. (LP: #426104)
  * Support for old Blogger 1.0 API.
  * i18n: Better support for internationalization.
  * Offline editing support. Use File/Go Offline to edit without internet
    connection.

 -- Jay Reding <wombat@serenity>  Mon, 21 Sep 2009 22:12:32 -0500

blogtk (2.0~ppa2~jaunty) jaunty; urgency=low

  * Fixed link insertion dialog bug

 -- Jay Reding <jay@jayreding.com>  Sun, 13 Sep 2009 00:17:09 -0500

blogtk (2.0~ppa1~jaunty) jaunty; urgency=low

  * Complete rewrite of BloGTK codebase.

 -- Jay Reding <jay@jayreding.com>  Sat, 12 Sep 2009 11:54:24 -0500

blogtk (1.2~ppa1~karmic) karmic; urgency=low
   * Appled patches from Ubuntu Jaunty repository
   * Deprecated file selection patch from Alessio Treglia (LP: #271311)
   * ConfigParser fix from Ralph Taylor (LP: #364745)

 -- Jay Reding <jay@jayreding.com>  Mon, 03 Aug 2009 11:27:11 +0100

blogtk (1.1-2ubuntu3) karmic; urgency=low

  * debian/patches/zz-fix-deprecated-configparser-options.patch (LP: #364745)
     - Python 2.6 Rev 60976 has deprecated usage of the "Default" section
       in ConfigParser.
     - Updated source to use "Default Server" instead of "Default".
     - Patch file has conflict with earlier non numbered patches so prefixed 
       with "zz" to ensure it is applied last.

 -- Ralph Taylor <djnrrd@gmail.com>  Sun, 21 Jun 2009 13:58:14 +0100

blogtk (1.1-2ubuntu2) hardy; urgency=low

  * debian/install:
     - Add three missing files (b-32.png, blogtk-icon.png, blogtk.glade).
  * debian/patches/fix-deprecated-true-false.patch:
     - Replace Gtk.TRUE with True and GTK.False with False in all
       Python files to avoid a large amount of deprecation warnings.
  * debian/BloGTK, debian/patches/hide-abort-traceback.patch:
     - Catch KeyboardInterrupt exceptions to hide ugly tracebacks.

 -- Siegfried-Angel Gevatter Pujals (RainCT) <rainct@ubuntu.com>  Fri, 29 Feb 2008 23:35:23 +0100

blogtk (1.1-2ubuntu1) hardy; urgency=low

  [ Siegfried-Angel Gevatter Pujals (RainCT) ]
  * Complete package clean-up (this package is now only maintained in
    Ubuntu, it was removed from Debian).
  * debian/control:
     - Change Maintainer field to Ubuntu MOTU Developers.
     - Drop docbook-to-man build dependency, as the manpage has been
       replaced by a new one written in groff (see below).
     - Drop imagemagick build dependecy, as we don't install a 16x16
       version of the menu anymore.
     - Add cdbs build dependency.
     - Move debhelper from Build-Depends-Indep to Build-Depends.
     - Bump standards version to 3.7.3 and add a Homepage field.
     - Fix XS-Python-Version field.
     - Add Vcs-Bzr and Vcs-Broswer fields.
  * debian/rules:
     - Switch to cdbs and remove all now unneeded stuff.
     - Add simple-patchsys.
     - Remove the 16x16 version of the menu.
  * debian/pycompat:
     - Remove it, as it isn't necessary with pycentral.
  * debian/blogtk.menu:
     - Change section to the new "Applications/Network/Web News".
     - Formatting changes.
     - Rename it to debian/menu.
  * debian/blogtk.desktop:
     - Add a new .desktop file which respects FreeDesktop.org's
       standards and contains Catalan and Spanish translations.
  * debian/compat:
     - Bump to version 5 as it has some nice features which can't harm.
  * debian/blogtk.1, debian/blogtk.sgml, debian/manpages:
     - Remove the old manpage and add a new one written in groff.
     - Add a debian/manpages file to install it.
  * debian/copyright:
     - Change download URL to a more direct one.
     - Add licensing information for the new manpage.
     - Mention re-packaging.
     - Formatting changes.
  * debian/install:
     - Add this files and install the files with it instead of using
       Upstream's Makefile, which would need to be patched and is uglier.
  * debian/blogtk.links:
     - Rename this file to debian/links.
     - Change the executable's name to be all lowercase.
  * debian/BloGTK:
     - Add this launcher (which prints a warning message and the runs
       the applicaiton) for a clean transition to the lowercase executable.
  * debian/blogtk.docs, debian/blogtk.dirs:
     - Remove them.
  * debian/patches/fix-shabangs.patch:
     - Fix shabangs for all .py files in src/.
  * debian/patches/fix-python25-crash.patch:
     - Fix a crash with URLs that don't start with http/https with
       Python 2.5, thanks to Simon Ochsenreither!


share/blogtk2/lib/blogtk2/config.py
#!/usr/bin/env python

# Copyright 2009 Jay Reding

# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
# 	http://www.apache.org/licenses/LICENSE-2.0
# 
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

try:
  from xml.etree import ElementTree # for Python 2.5 users
except:
  from elementtree import ElementTree

import xml.dom.minidom

# This library in great for pulling data from even badly-formed
# markup
from BeautifulSoup import BeautifulSoup

import gtk
import gtk.glade
import gobject
import pango

import base64
import urllib2
import threading

import os

from blogtk2 import SHARED_FILES

# This dictionary matches the systems with each blogging API they use.
systems = {
	'WordPress' : 'mt',
	'Movable Type' : 'mt',
	'Melody' : 'mt',
	'TextPattern' : 'mt',
	'Windows Live Spaces' : 'metaweblog',
	'Blogger' : 'gdata',
	'MetaWeblog' : 'metaweblog',
	'b2evolution' : 'metaweblog',
	'Habari'	:	'metaweblog',
	'LifeType' : 'metaweblog',
	'Drupal'	:	'metaweblog',
	'Expression Engine' : 'metaweblog',
	'Blogger 1.0' : 'blogger'
}

def threaded(f):
    def wrapper(*args):
        t = threading.Thread(target=f, args=args)
        t.start()
    return wrapper


class ConfigReader:

	def __init__(self):
		self.configDir = os.path.expanduser('~') + '/.BloGTK'
		self.configFile = self.configDir + '/config.xml'
		self.tree = ElementTree.parse(self.configFile).getroot()

	def getConfigTree(self):

		return self.tree

	def getConfigArray(self):

		accountData = []

		blogs = self.tree.findall('blog')
		for blog in blogs:
			account = { 
				'name' : blog.attrib['name'],
				'id' : blog.attrib['id'],
				'uri' : blog.find('uri').text,
				'endpoint' : blog.find('endpoint').text,
				'api' : blog.find('api').text,
				'system' : blog.find('system').text,
				'username' : blog.find('username').text,
				'password' : base64.b64decode(blog.find('password').text),
				'pullcount' : int(blog.find('pullcount').text)
			}
			accountData.append(account)

		return accountData
		
	def getCategoryArray(self, blogname):

		# ElementTree can't handle a simple XPath query like blog[@name = "somename"] -- lame.
		
		blogs = self.tree.findall('blog')

		catArray = []

		for blog in blogs:
			if blog.attrib['name'] == blogname:
				cats = blog.findall('categories/category')
				# This is like herding cats
				for cat in cats:
					category = {
						'categoryId' : cat.attrib['id'],
						'categoryName' : cat.attrib['name']
					}
					catArray.append(category)

		return catArray

class ConfigWriter:

	def __init__(self):
		self.configFile = os.path.expanduser('~') + '/.BloGTK' + '/config.xml'
		self.tree = ElementTree.parse(self.configFile).getroot()

	def writeCategoryArray(self, blogname, catsArray):

		# We need to refresh our config file object each time we do a write.
		# TODO: Better config file handling.
		self.configFile = os.path.expanduser('~') + '/.BloGTK' + '/config.xml'
		self.tree = ElementTree.parse(self.configFile).getroot()
		
		blogs = self.tree.findall('blog')
		for blog in blogs:
			if blog.attrib['name'] == blogname:
				cats = blog.find('categories')
				if cats:
					blog.remove(cats)
				newCat = ElementTree.Element('categories')
				blog.append(newCat)
				for cat in catsArray:
					try:
						newCat.append(ElementTree.Element('category', name=cat['categoryName'], id=cat['categoryId']))
					except:
						# This is due to Windows Live Spaces not handling categories in
						# a rational manner. Quelle surprise.
						newCat.append(ElementTree.Element('category', name=cat['title'], id=cat['title']))

				ElementTree.ElementTree(self.tree).write(self.configFile)

				# For security, we want our config file set chmod 0600 so others can't read the file.
				os.chmod(self.configFile, 0600)

class ConfigGUI:

	def __init__(self, mainInstance):

		self.mainInstance = mainInstance

		self.glade = gtk.glade.XML(os.path.join(SHARED_FILES, 'glade', 'blogtk2.glade'))

		self.dlgAccounts = self.glade.get_widget('dlgAccounts')
		self.dlgAccounts.connect('delete_event', self.closeWindow)

		# Grab our widgets... which sounds faintly dirty...
		self.tvAccountList = self.glade.get_widget('tvAccountList')
		self.teAcctName = self.glade.get_widget('teAcctName')
		self.teAcctAddress = self.glade.get_widget('teAcctAddress')
		self.teAcctEndpoint = self.glade.get_widget('teAcctEndpoint')
		self.teAcctUser = self.glade.get_widget('teAcctUser')
		self.teAcctPass = self.glade.get_widget('teAcctPass')
		self.teAcctBlogID = self.glade.get_widget('teAcctBlogID')
		self.cmbAcctSystem = self.glade.get_widget('cmbAcctSystem')
		self.spinPullcount = self.glade.get_widget('spinPullcount')
		self.tvFontPreview = self.glade.get_widget('tvFontPreview')

		#Set up our TreeView
		self.tsAccts = gtk.TreeStore( gobject.TYPE_STRING )
		self.tvAccountList.set_model(self.tsAccts)
		self.column0 = gtk.TreeViewColumn("Account Name", gtk.CellRendererText(), text=0 )
		self.tvAccountList.append_column(self.column0)

		# Grab our initial config
		self.accountTree = ConfigReader().getConfigTree()

		# Initiate Blog System Listing
		self.initSysList()
	
		# Populate our accounts
		self.populateAccountList()

		# Set the font preview for our editor font
		self.setFontPreview()

		# Initialize our change flag. If there is a change to any data, we need to set this to true
		self.changeFlag = False

		# This prevents multiple name error problems.
		self.nameErrorFlag = False

		# This array will hold our deleted blogs. On exit, should any blogs be deleted, we need to
		# delete their cache file as well.
		self.deletedBlogArray = []

		self.dlgAccounts.show()

		dic = { 
			'on_tvAccountList_cursor_changed' : self.changeAcct,
			'on_btnAcctIntrospection_clicked' : self.doIntrospection,
			'on_btnAcctAdd_clicked' : self.addAccount,
			'on_btnAccountRemove_clicked' : self.removeAccount,
			'on_teAcctName_focus_out_event' : self.updateData,
			'on_teAcctAddress_focus_out_event' : self.updateData,
			'on_teAcctEndpoint_focus_out_event' : self.updateData,
			'on_teAcctUser_focus_out_event' : self.updateData,
			'on_teAcctPass_focus_out_event' : self.updateData,
			'on_teAcctBlogID_focus_out_event' : self.updateData,
			'on_cmbAcctSystem_changed' : self.updateData,
			'on_spinPullcount_focus_out_event' : self.updateData,
			'on_widgetChanged_event' : self.switchChangeFlag,
			'on_btnAccountAccept_clicked' : self.saveAccount,
			'on_btnEditorFont_clicked' : self.setEditorFont,
			'on_btnAccountCancel_clicked' : self.closeWindow
		}
		self.glade.signal_autoconnect(dic)

	def initSysList(self):

		self.liststore = gtk.ListStore(str, str, 'gboolean')

		self.cmbAcctSystem.set_model(self.liststore)

		self.tvcolumn = gtk.TreeViewColumn()

		for item in sorted(systems.keys()):
			self.liststore.append([0, item, True])

		self.cellpb = gtk.CellRendererPixbuf()
		self.cell = gtk.CellRendererText()

		self.cmbAcctSystem.pack_start(self.cellpb, False)
		self.cmbAcctSystem.pack_start(self.cell, True)


		#self.cmbAcctSystem.set_cell_data_func(self.cellpb, self.make_pb)
		self.cmbAcctSystem.set_attributes(self.cell, text=1)

	def populateAccountList(self, index=None):

		self.tsAccts.clear()

		blogs = self.accountTree.findall('blog')

		for blog in blogs:
			self.tsAccts.append(None, [blog.attrib['name']])

		if not index == None:
			self.tvAccountList.get_selection().select_path(index)	
			self.tvAccountList.set_cursor(index, None, False )

		if index == None:
			# Because predestination is cool...
			sel = self.tvAccountList.get_selection()
			sel.set_mode(gtk.SELECTION_SINGLE)

			sel.select_path(0)
			self.tvAccountList.set_cursor(0, None, False )

			self.changeAcct(None)

	def changeAcct(self, widget, data=None):
		
		selected_iter = self.tvAccountList.get_selection().get_selected()[1]
		acctName = self.tsAccts.get_value(selected_iter, 0)

		blogs = self.accountTree.findall('blog')

		for blog in blogs:
			if blog.attrib['name'] == acctName:
				self.teAcctName.set_text(blog.attrib['name'])
				self.teAcctAddress.set_text(blog.find('uri').text)
				self.teAcctEndpoint.set_text(blog.find('endpoint').text)
				self.teAcctUser.set_text(blog.find('username').text)
				self.teAcctPass.set_text(base64.b64decode(blog.find('password').text))
				self.teAcctBlogID.set_text(blog.attrib['id'])
				if blog.find('system').text == 'Blogger':
					self.spinPullcount.set_value(25)
					self.spinPullcount.set_sensitive(False)
				else:
					self.spinPullcount.set_sensitive(True)
					self.spinPullcount.set_value(float(blog.find('pullcount').text))

				self.activeBlog = blog

		match_iter = self.search(self.liststore, self.liststore.iter_children(None), self.match_func, (1, self.activeBlog.find('system').text))
		self.cmbAcctSystem.set_active_iter(match_iter)


	# ADD ACCOUNT FUNCTIONS
	# ==================================================

	def addAccount(self, widget, data=None):
			
		# We don't want to add a new account until the last one has
		# been given a new name

		blogs = self.accountTree.findall('blog')

		for blog in blogs:
			if blog.attrib['name'] == 'New Account':
				return False
		
		e = ElementTree.Element

		# When we add, change, or remove any of our blog settings, we're doing so to an
		# in-memory copy of the actual config XML tree. This makes things immensely
		# simpler.
		newAccount = e('blog', name='New Account', id='')
		endpoint = e('endpoint')
		endpoint.text = ''		
		newAccount.append(endpoint)
		uri = e('uri')
		uri.text = ''		
		newAccount.append(uri)
		username = e('username')
		username.text = ''		
		newAccount.append(username)
		password = e('password')
		password.text = ''		
		newAccount.append(password)
		api = e('api')
		api.text = ''		
		newAccount.append(api)
		system = e('system')
		system.text = 'Blogger'		
		newAccount.append(system)
		pullcount = e('pullcount')
		pullcount.text = '20'
		newAccount.append(pullcount)

		self.accountTree.append(newAccount)

		self.activeBlog = newAccount

		# We want to select the newly-created account.
		self.populateAccountList(len(self.accountTree) - 2)

		# Set our change flag to true
		self.changeFlag = True

		# Here the focus gets set to the account name field.
		self.teAcctName.grab_focus()

	def updateData(self, widget, data=None):

		if self.liststore.get_value(self.cmbAcctSystem.get_active_iter(), 1) == 'Blogger':
			self.spinPullcount.set_sensitive(False)
		else:
			self.spinPullcount.set_sensitive(True)

		if widget.name == 'teAcctName':
			x = self.teAcctName.get_text()
			acctName = unicode(x, 'utf-8')

			# We need to get the selection, since we don't want to compare
			# names with the currently selected item.
			selected_iter = self.tvAccountList.get_cursor()[0][0]

			names = []

			for index, row in enumerate(self.tsAccts):
				if index == selected_iter:
					pass
				else:
					if acctName == row[0]:
						if self.nameErrorFlag == False:
							dialog = self.dialogCreator('Accounts must have unique names')
							self.teAcctName.grab_focus()
							dialog.destroy()
						
							self.nameErrorFlag = True

						self.teAcctName.grab_focus()

						return False

			self.activeBlog.attrib['name'] = acctName

			self.populateAccountList(selected_iter)
		
		if widget.name == 'teAcctAddress':
			url = widget.get_text()
			if not (url.startswith("http://") or url.startswith("https://")):
				url = "http://" + url
			self.activeBlog.find('uri').text = url

		if widget.name == 'teAcctEndpoint':
			url = widget.get_text()
			if not (url.startswith("http://") or url.startswith("https://")):
				url = "http://" + url
			self.activeBlog.find('endpoint').text = url
			
		if widget.name == 'teAcctUser':
			self.activeBlog.find('username').text = widget.get_text()

		if widget.name == 'teAcctPass':
			self.activeBlog.find('password').text = base64.b64encode(widget.get_text())

		if widget.name == 'teAcctBlogID':
			self.activeBlog.attrib['id'] = widget.get_text()

		if widget.name == 'cmbAcctSystem':		
			system = self.liststore.get_value(self.cmbAcctSystem.get_active_iter(), 1)
			self.activeBlog.find('system').text = system

		if widget.name == 'spinPullcount':
			if self.liststore.get_value(self.cmbAcctSystem.get_active_iter(), 1) == 'Blogger':
				self.activeBlog.find('pullcount').text = '25'
			else:
				self.activeBlog.find('pullcount').text = str(self.spinPullcount.get_value())[:-2]

		# We need to set up our api to match our system
		system = self.liststore.get_value(self.cmbAcctSystem.get_active_iter(), 1)
		self.activeBlog.find('api').text = systems[system]

		# We also need to update our change flag
		self.changeFlag == True


	# REMOVE ACCOUNT FUNCTIONS
	# ==================================================

	def removeAccount(self, widget, data=None):

		if len(self.accountTree) > 1:

			self.deletedBlogArray.append(self.activeBlog)
	
			self.accountTree.remove(self.activeBlog)

			
		else:
			dialog = self.dialogCreator(_('Must have at least one account'))

			dialog.destroy()

		self.populateAccountList(len(self.accountTree) - 1)

	# INTROSPECTION FUNCTIONSs
	# ==================================================

	@threaded
	def doIntrospection(self, widget, data=None):
		
		url = self.teAcctAddress.get_text()
		if not (url.startswith("http://") or url.startswith("https://")):
			url = "http://" + url

		gtk.gdk.threads_enter()
		winProgress = gtk.Window()
		#winProgress.set_property('skip-taskbar-hint', True)
		winProgress.set_transient_for(self.dlgAccounts)
		winProgress.set_position(gtk.WIN_POS_CENTER_ON_PARENT)
		winProgress.set_type_hint(gtk.gdk.WINDOW_TYPE_HINT_UTILITY)
		winProgress.set_title(_('Retrieving Blog Settings...'))
		hBox = gtk.HBox(spacing=8)
		vBox = gtk.VBox(spacing=8)
		progBar = gtk.ProgressBar()
		progBar.set_text(_('Retrieving Blog Settings...'))
		vBox.pack_start(progBar, padding=8)
		winProgress.add(hBox)
		hBox.pack_start(vBox, padding=12)
		winProgress.set_modal(True)
		winProgress.set_decorated(False)
		winProgress.resize(300, 40)
		winProgress.show_all()
		timeout = gobject.timeout_add(100, self.update_progress_bar, progBar)
		gtk.gdk.threads_leave()

		try:
			name, endpoint, blogID = self.processRSD(url)
		except Exception, e:
			print Exception, str(e)

			winProgress.destroy()

			dialog = self.dialogCreator(_('Sorry, but BloGTK could not retrieve your settings. Please enter them manually.'))

			gobject.source_remove(timeout)
			dialog.destroy()

			return False

		# Special processing for Blogspot and Blogger. The endpoint should equal the URL for the GData API.
		# Blogger's RSD points to the old API.
		if endpoint == 'http://www.blogger.com/api':
			endpoint = url

		# Well also need to know if we're dealing with Windows Live Spaces
		if endpoint == 'https://storage.msn.com/storageservice/MetaWeblog.rpc':
			name = 'Windows Live Spaces'

		# MovableType incorrectly prefers MetaWeblog to the MovableType API (!)
		# We should correct this
		if endpoint[-13:] == 'mt-xmlrpc.cgi':
			name = 'Movable Type'

		# We also need to correct for different spellings of Movable Type
		if name == 'MovableType':
			name = 'Movable Type'

		match_iter = self.search(self.liststore, self.liststore.iter_children(None), self.match_func, (1, name))
		self.cmbAcctSystem.set_active_iter(match_iter)
		self.teAcctEndpoint.set_text(endpoint)
		self.teAcctBlogID.set_text(blogID)

		# Here's where we also set our variables in the config object
		self.activeBlog.find('system').text = name
		self.activeBlog.attrib['id'] = blogID
		self.activeBlog.find('endpoint').text = endpoint

		gobject.source_remove(timeout)
		winProgress.destroy()
	
	def getIntrospectionURI(self, url):

		f = urllib2.urlopen(url)
		s = f.read()
		
		soup = BeautifulSoup(s)
		
		# If the template doesn't have RSD autodiscovery, we'll guess at
		# the RSD file location
		try:		
			rsdurl = soup.findAll(rel='EditURI')[0]['href']
		except:
			if url.endswith('/'):
				rsdurl = url + 'rsd.xml'
			else:
				rsdurl = url + '/rsd.xml'
		
		return rsdurl
		
	def processRSD(self, url):
	  
		rsdurl = self.getIntrospectionURI(url)
		
		f = urllib2.urlopen(rsdurl)
		s = f.read()
		
		doc = xml.dom.minidom.parseString(s)
		
		for api in doc.getElementsByTagName('api'):
		   if api.getAttribute('preferred') == 'true':
		      name = api.getAttribute('name')
		      endpoint = api.getAttribute('apiLink')
		      blogID = api.getAttribute('blogID')
		   else:
		      pass
		
		return name, endpoint, blogID

	def update_progress_bar(self, progBar):
		progBar.pulse()
		return True
		

	# ACCOUNT SAVING FUNCTIONS
	# ================================================== 

	def saveAccount(self, widget, data=None):

		# We need to loop through our accounts and make sure everything is filled.
		# TODO Throw a dialog here if there's an error. Also preselect the blog with the missing info

		for blog in self.accountTree.findall('blog'):
			if blog.attrib['name'] == '':
				dialog = self.dialogCreator(_('Must name your blog'))
				dialog.destroy()
				return False
			if blog.find('uri').text == '':
				dialog = self.dialogCreator(_('Must have a URI'))
				dialog.destroy()
				return False
			if blog.find('endpoint').text == '':
				dialog = self.dialogCreator(_('Must have an endpoint'))
				dialog.destroy()
				return False
			if blog.find('username').text == '':
				dialog = self.dialogCreator(_('Must have a username'))
				dialog.destroy()
				return False
			if blog.find('password').text == '':
				dialog = self.dialogCreator(_('Must have a password'))
				dialog.destroy()
				return False

		# We also need to loop through our deleted blogs array and remove any remaining cache files.
		for blog in self.deletedBlogArray:

			accountID = blog.find('endpoint').text + '/' +  blog.attrib['id']

			cacheFile = ConfigReader().configDir + '/cache/' + base64.b64encode(accountID) + ".xml"

			try:
				os.remove(cacheFile)
			except Exception, e:
				print Exception, str(e)
				

		ElementTree.ElementTree(self.accountTree).write(ConfigReader().configFile)

		# For security, we want our config file set chmod 0600 so others can't read the file.
		os.chmod(ConfigReader().configFile, 0600)

		self.mainInstance.refreshBlogList()

		self.dlgAccounts.destroy()

	# SYSTEM SETTINGS FUNCTIONS
	# ==================================================

	def setFontPreview(self, widget=None, data=None):
		

		settings = self.accountTree.find('settings')
		fontname = settings.find('editorfont').text
		self.tvFontPreview.get_buffer().set_text(fontname)
		fontdesc =  pango.FontDescription(fontname)
		self.tvFontPreview.modify_font(fontdesc)
		

	def setEditorFont(self, widget, data=None):

		settings = self.accountTree.find('settings')

		dlgSelectFont = gtk.FontSelectionDialog(_('Select Editor Font'))
		dlgSelectFont.set_icon(gtk.gdk.pixbuf_new_from_file(os.path.join(SHARED_FILES, 'res', 'b-32.png')))
		dlgSelectFont.set_transient_for(self.dlgAccounts)
		dlgSelectFont.set_font_name(settings.find('editorfont').text)
		result = dlgSelectFont.run()
		if result == gtk.RESPONSE_OK:
			fontname = dlgSelectFont.get_font_name()

			# Set our preview to display the selected font
			self.tvFontPreview.get_buffer().set_text(fontname)
			fontdesc =  pango.FontDescription(fontname)
			self.tvFontPreview.modify_font(fontdesc)

			settings.find('editorfont').text = fontname

			dlgSelectFont.destroy()

		else:
			dlgSelectFont.destroy()

	def closeWindow(self, widget, data=None):
		if self.changeFlag == True:
			msgbox = gtk.MessageDialog(parent = self.dlgAccounts, buttons = gtk.BUTTONS_OK_CANCEL, flags = gtk.DIALOG_MODAL, type = gtk.MESSAGE_WARNING, message_format = _('Are you sure you wish to close this window?'))
			msgbox.set_title("Confirm Close")
			msgbox.format_secondary_text(_('All unsaved changes will be lost.'))
			result = msgbox.run()
			msgbox.destroy()
			if result == gtk.RESPONSE_OK:
				self.dlgAccounts.destroy()
				return True
			else:
				return False
		else:
			self.dlgAccounts.destroy()
			return True

	def switchChangeFlag(self, widget, data=None):
		self.changeFlag = True


	# CONVENIENCE FUNCTIONS
	# ==================================================

	def make_pb(self, tvcolumn, cell, model, iter):
		stock = model.get_value(iter, 0)
		pb = self.cmbAcctSystem.render_icon(stock, gtk.ICON_SIZE_MENU, None)
		cell.set_property('pixbuf', pb)
		return

	def match_func(self, model, iter, data):
		column, key = data # data is a tuple containing column number, key
		value = model.get_value(iter, column)
		return value == key

	def search(self, model, iter, func, data):
		while iter:
			if func(model, iter, data):
				return iter
			result = self.search(model, model.iter_children(iter), func, data)
			if result: return result
			iter = model.iter_next(iter)
		return None
 
	def dialogCreator(self, message):
		dialog = gtk.MessageDialog(parent=self.dlgAccounts, flags=0, type=gtk.MESSAGE_ERROR, buttons=gtk.BUTTONS_OK, message_format=message)
		dialog.run()

		return dialog
